### DISCUSSION TITLE
"Where is the source, Luke?"

### DISCUSSION THREAD

**Discussion by mcctuxic** [tabemann/zeptoforth#115] [General]
Hi,

Do you know of a good source for forth definitions of the basic words - those zeptoforth as implemented in assembly language?

Do you know an Android USB terminal app, which works with zeptoforth? For "Zeptoforth on the go"... :)

Cheers!
Tuxic

**Comment by tabemann** [comment]
On Wed, Jul 17, 2024 at 12:49 PM mcctuxic ***@***.***> wrote:

> Hi,
>
> Do you know of a good source for forth definitions of the basic words -
> those zeptoforth as implemented in assembly language?
>

Those were hand-coded in assembly and do not have Forth definitions. Also,
they do things with registers that zeptoforth itself is not smart enough to
do (as zeptoforth works primarily with R6 for the top of the stack and R7
for the address of the top of the data stack, and R0-R3 (and to a lesser
extent R4 and R5) are used merely as scratchpad space, while the code
written in assembly uses R0-R5 much more intensively than code written in
Forth.


> Do you know an Android USB terminal app, which works with zeptoforth? For
> "Zeptoforth on the go"... :)
>
> Cheers!
> Tuxic
>

I have never used a terminal program on my Android phone or tablets, so I
wouldn't know.

Travis

**Reply by mcctuxic** [reply]
Hi Travis,

sorry...I confused what wanted to say:
My question is completly unrelated to zeptoforth. 
I am just curious how all these "words, which were given by the masters" :) would be implemented in Forth and 
how many words are SO basic, that they can't be implemented in forth. A little bit like "how would be a computer like,
which has been bootstrapped from NAND gates alone"... kind of.

I the meanwhile I tried some terminal programs:
They fail either by not supporting ANSI, or being to old and could not handle 
the newer access rights to usb devices for Android >= 11.
The one closest to work spit a row of alien control codes which each OK at me.

 Cheers!
Tuxic

**Comment by tabemann** [comment]
On Sun, Jul 21, 2024, 09:40 mcctuxic ***@***.***> wrote:

> Hi Travis,
>
> sorry...I confused what wanted to say:
> My question is completly unrelated to zeptoforth.
> I am just curious how all these "words, which were given by the masters"
> :) would be implemented in Forth and
> how many words are SO basic, that they can't be implemented in forth. A
> little bit like "how would be a computer like,
> which has been bootstrapped from NAND gates alone"... kind of.
>
The number of words that must be primitives can be very small - look at
eForth for instance. However this kills performance compared to
implementing a greater number of words in assembly.

> I the meanwhile I tried some terminal programs:
> They fail either by not supporting ANSI, or being to old and could not
> handle
> the newer access rights to usb devices for Android >= 11.
> The one closest to work spit a row of alien control codes which each OK at
> me.
>
To get rid of these codes execute:

false ack-nak-enabled !
false bel-enabled !
false color-enabled !
disable-line

Note that these turn off ACK/NAK terminal synchronization, BEL on errors,
color on errors, and the line editor (if you have it enabled).

Travis

>
